crates/kornia-tensor-ops/src/ops.rs

```diff
@@ -76,6 +76,170 @@ where
     })
 }
 
+/// Compute the dot product between two tensors
+///
+/// # Arguments
+///
+/// * `a` - First tensor
+/// * `b` - Second tensor
+///
+/// # Returns
+///
+/// A scalar value representing the dot product of the two tensors.
+///
+/// # Errors
+///
+/// If the shapes of the tensors don't match, an error is returned.
+///
+/// # Example
+///
+/// ```
+/// use kornia_tensor::{Tensor, CpuAllocator};
+/// use kornia_tensor_ops::ops::dot_product;
+///
+/// let a = Tensor::<i32, 1, CpuAllocator>::from_shape_slice([3], &[1, 2, 3], CpuAllocator).unwrap();
+/// let b = Tensor::<i32, 1, CpuAllocator>::from_shape_slice([3], &[4, 5, 6], CpuAllocator).unwrap();
+/// let result = dot_product(&a, &b).unwrap();
+/// assert_eq!(result, 32); // 1*4 + 2*5 + 3*6 = 4 + 10 + 18 = 32
+/// ```
+pub fn dot_product<T, const N: usize, A>(
+    a: &Tensor<T, N, A>,
+    b: &Tensor<T, N, A>,
+) -> Result<T, TensorOpsError>
+where
+    T: Zero + Clone + std::ops::Add<Output = T> + std::ops::Mul<Output = T>,
+    A: TensorAllocator + Clone + 'static,
+{
+    if a.shape != b.shape {
+        return Err(TensorOpsError::ShapeMismatch(
+            a.shape.to_vec(),
+            b.shape.to_vec(),
+        ));
+    }
+
+    let mut result = T::zero();
+    for (a_val, b_val) in a.as_slice().iter().zip(b.as_slice().iter()) {
+        result = result + a_val.clone() * b_val.clone();
+    }
+
+    Ok(result)
+}
+
+/// Compute the cosine similarity between two tensors
+///
+/// # Arguments
+///
+/// * `a` - First tensor
+/// * `b` - Second tensor
+///
+/// # Returns
+///
+/// A scalar value representing the cosine similarity between the two tensors.
+/// The value ranges from -1 to 1, where:
+/// - 1 means the tensors are identical in direction (maximum similarity)
+/// - 0 means the tensors are orthogonal (no similarity)
+/// - -1 means the tensors are opposite in direction (maximum dissimilarity)
+///
+/// # Errors
+///
+/// If the shapes of the tensors don't match, an error is returned.
+///
+/// # Example
+///
+/// ```
+/// use kornia_tensor::{Tensor, CpuAllocator};
+/// use kornia_tensor_ops::ops::cosine_similarity;
+///
+/// let a = Tensor::<f32, 1, CpuAllocator>::from_shape_slice([3], &[1.0, 0.0, 0.0], CpuAllocator).unwrap();
+/// let b = Tensor::<f32, 1, CpuAllocator>::from_shape_slice([3], &[0.0, 1.0, 0.0], CpuAllocator).unwrap();
+/// let result = cosine_similarity(&a, &b).unwrap();
+/// assert_eq!(result, 0.0); // Orthogonal vectors have cosine similarity of 0
+/// ```
+pub fn cosine_similarity<T, const N: usize, A>(
```

Comment:
Can you check this post ? https://ashvardanian.com/posts/python-c-assembly-comparison/
there some ideas for optimisations here. Besides, can you write a small benchmark scripts (check in other crates the benches). Would be great to see performance number and make it close the ones in the post